### PR TITLE
Upgrade command

### PR DESCRIPTION
**Pull request by patrick91**
This PR introduces an upgrade command, it works like this:

```
strawberry upgrade annotated-union .
```

In future we'll allow to run all codemods at once 😊

### CODE REVIEW

File: strawberry/cli/__init__.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
-from .commands.codegen import codegen  # noqa
-from .commands.export_schema import export_schema  # noqa
-from .commands.server import server  # noqa
-
+from .commands.codegen import codegen as codegen  # noqa
```
Comment (patrick91) [inline strawberry/cli/__init__.py:1]: added `as ...` to make pyright happier 😊

File: strawberry/cli/commands/upgrade/__init__.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+from ._run_codemod import run_codemod
+
+codemods = {
+    "annotated-union": ConvertUnionToAnnotatedUnion,
+}
+
+
+# TODO: add support for running all of them
+# TODO: add support for passing a list of files
```
Comment (patrick91) [inline strawberry/cli/commands/upgrade/__init__.py:19]: this is mostly done, needs the ability of passing multiple paths

**Comment by botberry** [comment]
## Apollo Federation Subgraph Compatibility Results

<table>
<thead>
<tr><th>Federation 1 Support</th><th>Federation 2 Support</th></tr>
</thead>
<tbody>
<tr><td><table><tr><th><code>_service</code></th><td>🟢</td></tr><tr><th><code>@key (single)</code></th><td>🟢</td></tr><tr><th><code>@key (multi)</code></th><td>🟢</td></tr><tr><th><code>@key (composite)</code></th><td>🟢</td></tr><tr><th><code>repeatable @key</code></th><td>🟢</td></tr><tr><th><code>@requires</code></th><td>🟢</td></tr><tr><th><code>@provides</code></th><td>🟢</td></tr><tr><th><code>federated tracing</code></th><td>🔲</td></tr></table></td><td><table><tr><th><code>@link</code></th><td>🟢</td></tr><tr><th><code>@shareable</code></th><td>🟢</td></tr><tr><th><code>@tag</code></th><td>🟢</td></tr><tr><th><code>@override</code></th><td>🟢</td></tr><tr><th><code>@inaccessible</code></th><td>🟢</td></tr><tr><th><code>@composeDirective</code></th><td>🟢</td></tr><tr><th><code>@interfaceObject</code></th><td>🟢</td></tr></table></td></tr>
</tbody>
</table>

Learn more:
* [Apollo Federation Subgraph Specification](https://www.apollographql.com/docs/federation/subgraph-spec/)
* [Compatibility Tests](https://github.com/apollographql/apollo-federation-subgraph-compatibility/blob/main/COMPATIBILITY.md)

File: strawberry/cli/commands/upgrade/_run_codemod.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    from libcst.codemod import Codemod
+
+ProgressType = Union[type[Progress], type[FakeProgress]]
+PoolType = Union[type[Pool], type[DummyPool]]  # type: ignore
+
+
+def _execute_transform_wrap(
+    job: Dict[str, Any],
+) -> ExecutionResult:
+    # TODO: maybe capture warnings?
```
Comment (patrick91) [inline strawberry/cli/commands/upgrade/_run_codemod.py:24]: warnings happen when libcst failed to determine module name for the file, I was getting this a few times, not sure if it is important or not

File: strawberry/codemods/annotated_unions.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+class ConvertUnionToAnnotatedUnion(VisitorBasedCodemodCommand):
+    # TODO: support Union and | syntax, also in errors? ugh
+    DESCRIPTION: str = (
+        "Converts strawberry.union(..., types=(...)) to Annotated[Union[...]]"
+    )
+
+    def __init__(self, context: CodemodContext) -> None:
+        super().__init__(context)
+
+    @m.leave(m.Call(func=m.Attribute(value=m.Name("strawberry"), attr=m.Name("union"))))
```
Comment (Kludex) [inline strawberry/codemods/annotated_unions.py:39]: If you want to be more precise, you can check the:

```py
m.ImportFrom(
    import=m.Name("strawberry"),
    alias=[
        m.ZeroOrMore(),
        m.ImportAlias(value=cst.Name("union")),
        m.ZeroOrMore()
    ]
)
```
To see if `union` is imported from `strawberry`... And match with:

 ```suggestion
    @m.leave(m.Call(func=m.Attribute(value=m.Name("strawberry"), attr=m.Name("union")) | m.Name("union")))
```

File: strawberry/codemods/annotated_unions.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            return None
+
+        if index == search_index and arg.keyword is None:
+            return arg
+
+
+class ConvertUnionToAnnotatedUnion(VisitorBasedCodemodCommand):
+    # TODO: support Union and | syntax, also in errors? ugh
+    DESCRIPTION: str = (
+        "Converts strawberry.union(..., types=(...)) to Annotated[Union[...]]"
```
Comment (Kludex) [inline strawberry/codemods/annotated_unions.py:33]:
```suggestion
        "Converts strawberry.union(..., types=(...)) to Annotated[Union[...], strawberry.union(...)]"
```

File: tests/codemods/test_annotated_unions.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+            AUnion = strawberry.union(name="ABC", types=(Foo, Bar))
+        """
+
+        after = """
+            from typing_extensions import Annotated
+
+            AUnion = Annotated[Union[Foo, Bar], strawberry.union(name="ABC")]
+        """
+
+        self.assertCodemod(before, after)
```
Comment (Kludex) [inline tests/codemods/test_annotated_unions.py:24]:
```suggestion
        self.assertCodemod(before, after)
    
    def test_update_union_positional_name(self) -> None:
        before = """
            AUnion = strawberry.union("ABC", types=(Foo, Bar))
        """

        after = """
            from typing_extensions import Annotated
            AUnion = Annotated[Union[Foo, Bar], strawberry.union("ABC")]
        """

        self.assertCodemod(before, after)
```

File: strawberry/codemods/annotated_unions.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+
+    def __init__(self, context: CodemodContext) -> None:
+        super().__init__(context)
+
+    @m.leave(m.Call(func=m.Attribute(value=m.Name("strawberry"), attr=m.Name("union"))))
+    def leave_union_call(
+        self, original_node: Call, updated_node: Call
+    ) -> BaseExpression:
+        types = _find_named_argument(original_node.args, "types")
+        union_name = _find_named_argument(original_node.args, "name")
```
Comment (Kludex) [inline strawberry/codemods/annotated_unions.py:44]: There's a pattern that avoids this kind of logic, which is using a `m.visit`/`m.leave` with `m.call_if_inside`. In other words, you could `@m.leave(m.Arg())`, making sure you are inside a:

```py
m.Call(func=m.Attribute(value=m.Name("strawberry"), attr=m.Name("union")))
```

Comment (Kludex) [reply strawberry/codemods/annotated_unions.py:44]: Or you could visit first, store variables inside `self.<something>`, and then on `m.leave` you'd consume and reset them.

File: strawberry/codemods/annotated_unions.py (hunk 4 above)
Comment (patrick91) [reply strawberry/codemods/annotated_unions.py:39]: ah nice! thanks <3

File: strawberry/codemods/annotated_unions.py [hunk 8]
```diff
@@ -35,8 +35,28 @@ class ConvertUnionToAnnotatedUnion(VisitorBasedCodemodCommand):
...
+        self._is_using_named_import = False
+
+        return super().visit_Module(node)
+
+    @m.visit(
+        m.ImportFrom(
+            m.Name("strawberry"),
+            [
+                m.ZeroOrMore(),
+                m.ImportAlias(cst.Name("union")),  # type: ignore
```
Comment (Kludex) [inline strawberry/codemods/annotated_unions.py:52]:
```suggestion
                m.ImportAlias(m.Name("union")),
```

Comment (patrick91) [reply strawberry/codemods/annotated_unions.py]: hero

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

This release introduces a new command called `upgrade`, this command can be used
to run codemods on your codebase to upgrade to the latest version of Strawberry.

At the moment we only support upgrading unions to use the new syntax with
annotated, but in future we plan to add more commands to help with upgrading.

Here's how you can use the command to upgrade your codebase:

```shell
strawberry upgrade annotated-union .
```


---
Here's the preview release card for twitter:




Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to @patrick91 for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

**Pull request merged**